pkg/db-repo/change_history_manager.go

```diff
@@ -0,0 +1,232 @@
+package db_repo
+
+import (
+	"fmt"
+	"github.com/applike/gosoline/pkg/cfg"
+	"github.com/applike/gosoline/pkg/mon"
+	"github.com/jinzhu/gorm"
+	"reflect"
+	"strings"
+)
+
+type changeHistoryManagerSettings struct {
+	ChangeAuthorField string `cfg:"change_author_column"`
+	TableSuffix       string `cfg:"table_suffix" default:"history"`
+}
+
+type changeHistoryManager struct {
+	orm           *gorm.DB
+	logger        mon.Logger
+	settings      *changeHistoryManagerSettings
+	model         ModelBased
+	originalTable *tableMetadata
+	historyTable  *tableMetadata
+	statements    []string
+}
+
+func newChangeHistoryManager(config cfg.Config, logger mon.Logger, model ModelBased) *changeHistoryManager {
+	orm := NewOrm(config, logger)
+
+	settings := &changeHistoryManagerSettings{}
+	config.UnmarshalKey("change_history", settings)
+
+	return newChangeHistoryManagerWithInterfaces(logger, orm, model, settings)
+}
+
+func newChangeHistoryManagerWithInterfaces(logger mon.Logger, orm *gorm.DB, model ModelBased, settings *changeHistoryManagerSettings) *changeHistoryManager {
+	statements := make([]string, 0)
+
+	logger = logger.WithChannel("change_history_manager").WithFields(mon.Fields{
+		"model": reflect.TypeOf(model).Elem().Name(),
+	})
+
+	return &changeHistoryManager{
+		logger:     logger,
+		orm:        orm,
+		model:      model,
+		statements: statements,
+		settings:   settings,
+	}
+}
+
+func (c *changeHistoryManager) runMigration() error {
+	c.buildOriginalTableMetadata()
+	c.buildHistoryTableMetadata()
+
+	if !c.historyTable.exists {
+		c.createHistoryTable()
+		c.dropHistoryTriggers()
+		c.createHistoryTriggers()
+		c.logger.Info("creating change history setup")
+		return c.execute()
```

Comment:
i would extract this whole Statement generation in a seperate file. just to keep ist more readable.

Reply:
Interesting idea. I'll give it a try.